### PR TITLE
Disable beatmap skinning support when entering the skin editor

### LINKED ISSUE #22561: Skin editor overwrites gameplay layout after being opened during map with a legacy beatmap skin
### Type

Game behaviour

### Bug description

1. Have a triangles / argon skin that you applied customisations to
2. Enter gameplay in a map that has a legacy beatmap skin
3. Open skin editor and close it
4. Exit gameplay
5. "Skin saved" toast will show
6. Re-enter gameplay on map without beatmap skin
7. Original skin layout will be mangled

### Screenshots or videos

https://user-images.githubusercontent.com/20418176/217614365-855c6fd8-e4cd-44ab-90c7-ff44d71b7239.mp4

### Version

current master

### Logs

\-

### LINKED ISSUE #23915: Skin editor change will revert and not save in gameplay if change some specific visual settings
### Type

Game behaviour

### Bug description

1. Do some skin layout change in Skin Editor at Gameplay.
2. Change Beatmap skins or Beatamp colours in Visual Settings (Actually also work in settings layout which open by ctrl+O)
3. Change reverted.
4. Exit Gameplay and found change not saved.

Under normal circumstances, the skin will be saved when exiting player screen even didn't click save in Skin Editor.

### Screenshots or videos

https://github.com/ppy/osu/assets/34775378/295d08d8-2471-4975-b4bc-2b3df6f7056e

### Version

2023.614.1

### Logs

eh... no related log, just some `OsuScreenStack` log

```
2023-06-14 12:32:53 [verbose]: 📺 OsuScreenStack#527(depth:6) exit from ReplayPlayer#446
2023-06-14 12:32:53 [verbose]: 📺 OsuScreenStack#527(depth:6) resume to ReplayPlayerLoader#254
2023-06-14 12:32:53 [verbose]: 📺 OsuScreenStack#527(depth:5) exit from ReplayPlayerLoader#254
2023-06-14 12:32:53 [verbose]: 📺 OsuScreenStack#527(depth:5) resume to PlaySongSelect#317
```

### PR DESCRIPTION
**Pull request by peppy**
As proposed in https://github.com/ppy/osu/issues/23915.

Closes #23915 
Closes #22561

This *definitely* needs test coverage, but wanted to get this out to make sure people agree on the behaviour and direction before considering adding them.

### CODE REVIEW

File: osu.Game/Overlays/SkinEditor/SkinEditorOverlay.cs [hunk 1]
```diff
@@ -147,17 +156,24 @@ public void OnReleased(KeyBindingReleaseEvent<GlobalAction> e)
...
 
-            // AddOnce with parameter will ensure the newest target is loaded if there is any overlap.
-            Scheduler.AddOnce(setTarget, screen);
+                // AddOnce with parameter will ensure the newest target is loaded if there is any overlap.
+                Scheduler.AddOnce(setTarget, screen);
+            }
+            finally
+            {
+                globallyReenableBeatmapSkinSetting();
+            }
```
Comment (bdach) [inline osu.Game/Overlays/SkinEditor/SkinEditorOverlay.cs:176]: So `setTarget()` - which internally contains a `globallyDisableBeatmapSkinSetting()` call if the target screen is a `Player` - is scheduled, but then `globallyReenableBeatmapSkinSetting()` is called in a `finally` block? It's probably not *wrong*, as the schedule saves this, but why is this written like that in the first place? I assume the `finally` is aimed at the early-return, but if so, then can the re-enable not be done at the start of the method or something?

File: osu.Game/Overlays/SkinEditor/SkinEditorOverlay.cs [hunk 2]
```diff
@@ -182,5 +201,30 @@ private void setTarget(OsuScreen? target)
...
+        {
+            if (beatmapSkins.Disabled)
+                return;
+
+            // The skin editor doesn't work well if beatmap skins are being applied to the player screen.
+            // To keep things simple, disable the setting game-wide while using the skin editor.
+            beatmapSkinsOriginalState = beatmapSkins.Value;
+
+            beatmapSkins.Value = false;
+            beatmapSkins.Disabled = true;
```
Comment (bdach) [inline osu.Game/Overlays/SkinEditor/SkinEditorOverlay.cs:218]: Is there a reason why this is essentially manually reimplementing what a bindable lease is?

```diff
diff --git a/osu.Game/Overlays/SkinEditor/SkinEditorOverlay.cs b/osu.Game/Overlays/SkinEditor/SkinEditorOverlay.cs
index b120faa45f..4f0028de64 100644
--- a/osu.Game/Overlays/SkinEditor/SkinEditorOverlay.cs
+++ b/osu.Game/Overlays/SkinEditor/SkinEditorOverlay.cs
@@ -203,7 +203,7 @@ private void setTarget(OsuScreen? target)
         }
 
         private readonly Bindable<bool> beatmapSkins = new Bindable<bool>();
-        private bool beatmapSkinsOriginalState;
+        private LeasedBindable<bool>? leasedBeatmapSkins;
 
         private void globallyDisableBeatmapSkinSetting()
         {
@@ -212,19 +212,14 @@ private void globallyDisableBeatmapSkinSetting()
 
             // The skin editor doesn't work well if beatmap skins are being applied to the player screen.
             // To keep things simple, disable the setting game-wide while using the skin editor.
-            beatmapSkinsOriginalState = beatmapSkins.Value;
-
-            beatmapSkins.Value = false;
-            beatmapSkins.Disabled = true;
+            leasedBeatmapSkins = beatmapSkins.BeginLease(true);
+            leasedBeatmapSkins.Value = false;
         }
 
         private void globallyReenableBeatmapSkinSetting()
         {
-            if (!beatmapSkins.Disabled)
-                return;
-
-            beatmapSkins.Disabled = false;
-            beatmapSkins.Value = beatmapSkinsOriginalState;
+            leasedBeatmapSkins?.Return();
+            leasedBeatmapSkins = null;
         }
     }
 }

```

**Comment by bdach** [comment]
Additionally, if the skin editor is opened and exited on a beatmap with beatmap skins active, the skin editor will turn off beatmap skins and disable the toggle on open, but it will not re-enable and restore the previous value on exit:

https://github.com/ppy/osu/assets/20418176/9c40ea53-de42-462b-98ce-5b6a76d47d68

which I would say is bad.

**Comment by peppy** [comment]
> Additionally, if the skin editor is opened and exited on a beatmap with beatmap skins active, the skin editor will turn off beatmap skins and disable the toggle on open, but it will not re-enable and restore the previous value on exit:
> 
> 
> 
> https://github.com/ppy/osu/assets/20418176/9c40ea53-de42-462b-98ce-5b6a76d47d68
> 
> 
> 
> which I would say is bad.

Are you expecting it to re-enable without exiting player? Because that wasn't my intention with implementation (would require the skin editor to save changes in toggle rather than screen change).

**Comment by bdach** [comment]
> Are you expecting it to re-enable without exiting player? Because that wasn't my intention with implementation (would require the skin editor to save changes in toggle rather than screen change).

I suppose so. It feels kinda janky that this entirely unrelated action of opening the skin editor causes beatmap skins to be disabled (for good reasons) but then doesn't put things back as they were once you're done editing your personal skin. Not sure.

**Comment by peppy** [comment]
I guess we can go in that direction. What do you think about providing confirmation to save when exiting the editor in that case? Should it be something we expedite since it's happening a lot more often?

**Comment by bdach** [comment]
> What do you think about providing confirmation to save when exiting the editor in that case?

I kinda think that's not really needed. It can just silently save the same way the skin editor already does on screen change.

**Comment by peppy** [comment]
Okay, I'll give that a try. It may be okay now that saving skins is a lot faster (in the past there was a significant overhead).

**Comment by peppy** [comment]
I've made the requested changes. Feels pretty good IMO.

**Review by bdach — APPROVED** [review]

**Pull request merged**